go-controller/pkg/ovn/gateway_init.go

```diff
@@ -316,91 +316,132 @@ func addDistributedGWPort() error {
-		fmt.Sprintf("name=%s_%s", dgpName, masterChassisID), "priority=100",
-		"--", "set", "logical_router_port", dgpName, "gateway_chassis=@gw")
-	stdout, stderr, err := util.RunOVNNbctl(nbctlArgs...)
-	if err != nil {
-		return fmt.Errorf("failed to set gateway chassis %s for distributed gateway port %s: "+
-			"stdout: %q, stderr: %q, error: %v", masterChassisID, dgpName, stdout, stderr, err)
-	}
-
-	// connect the distributed gateway port to logical switch configured with localnet port
-	nbctlArgs = []string{
-		"--may-exist", "ls-add", nodeLocalSwitch,
-	}
-	// add localnet port to the logical switch
-	lclNetPortname := "lnet-" + nodeLocalSwitch
-	nbctlArgs = append(nbctlArgs,
-		"--", "--may-exist", "lsp-add", nodeLocalSwitch, lclNetPortname,
-		"--", "set", "logical_switch_port", lclNetPortname, "addresses=unknown", "type=localnet",
-		"options:network_name="+util.LocalNetworkName)
-	// connect the switch to the distributed router
-	lspName := switchToRouterPrefix + nodeLocalSwitch
-	nbctlArgs = append(nbctlArgs,
-		"--", "--may-exist", "lsp-add", nodeLocalSwitch, lspName,
-		"--", "set", "logical_switch_port", lspName, "type=router", "addresses=router",
-		"options:nat-addresses=router", "options:router-port="+dgpName)
-	stdout, stderr, err = util.RunOVNNbctl(nbctlArgs...)
-	if err != nil {
-		return fmt.Errorf("failed creating logical switch %s and its ports (%s, %s) "+
-			"stdout: %q, stderr: %q, error: %v", nodeLocalSwitch, lclNetPortname, lspName, stdout, stderr, err)
-	}
-	// finally add an entry to the OVN SB MAC_Binding table, if not present, to capture the
-	// MAC-IP binding of util.V4NodeLocalNatSubnetNextHop address to its MAC. Normally, this will
-	// be learnt and added by the chassis to which the distributed gateway port (DGP) is
-	// bound. However, in our case we don't send any traffic out with the DGP port's IP
-	// as source IP, so that binding will never be learnt and we need to seed it.
-	dnatSnatNextHopMac := util.IPAddrToHWAddr(net.ParseIP(util.V4NodeLocalNatSubnetNextHop)).String()
-	stdout, stderr, err = util.RunOVNSbctl("--data=bare", "--no-heading", "--columns=_uuid", "find", "MAC_Binding",
-		"logical_port="+dgpName, fmt.Sprintf(`mac="%s"`, dnatSnatNextHopMac))
-	if err != nil {
-		return fmt.Errorf("failed to check existence of MAC_Binding entry of (%s, %s) for distributed router port %s "+
-			"stderr: %q, error: %v", util.V4NodeLocalNatSubnetNextHop, dnatSnatNextHopMac, dgpName, stderr, err)
-	}
-	if stdout != "" {
-		klog.Infof("The MAC_Binding entry of (%s, %s) exists on distributed router port %s with uuid %s",
-			util.V4NodeLocalNatSubnetNextHop, dnatSnatNextHopMac, dgpName, stdout)
-		return nil
-	}
+	var ipModes []string
+	if config.IPv4Mode {
+		ipModes = append(ipModes, "IPv4Mode")
+	}
+	if config.IPv6Mode {
+		ipModes = append(ipModes, "IPv6Mode")
+	}
+
+	var macStr string
+	for _, ipMode := range ipModes {
```

Comment:
hm... this is weird. I think I suggested using a loop before but I was thinking you could loop over something that already existed (`subnets` or `ips` or something).

If not maybe it would be clearer to split the per-family setup code out into a separate function and then do

    if config.IPv4Mode {
            err := setupOneWhatever(util.V4NodeLocalDistributedGwPortIP, util.V4NodeLocalNatSubnetNextHop)
            if err != nil { ... }
    }
    if config.IPv6Mode {
            err := setupOneWhatever(util.V6NodeLocalDistributedGwPortIP, util.V6NodeLocalNatSubnetNextHop)
            if err != nil { ... }
    }